### DISCUSSION TITLE
25. HTTP 메서드와 이것이 하는 역할에 대해 설명해 주세요.

### DISCUSSION THREAD

**Discussion by cutehumanS2** [jewelry-box/cs-interview#6] [General]
HTTP 메서드와 이것이 하는 역할에 대해 설명해 주세요.

**Comment by linirini** [comment]
HTTP 메서드는 클라이언트가 서버에 요청을 보낼 때 사용하는 동작의 종류를 나타낸다. HTTP 프로토콜은 다양한 메서드를 제공하며 ,각각의 메서드는 특정한 작업을 수행하도록 설계되었다.
HTTP 메서드의 속성으로 안전, 멱등, 캐시 가능이 있다.
- 안전성: HTTP 메서드가 서버의 상태를 변경하지 않는지를 나타내는 속성
- 멱등성: 동일한 요청을 여러번 보냈을 때 서버 상태가 동일하게 유지되는지를 나타내는 속성
- 캐시 가능: HTTP 응답이 클라이언트 또는 중간 프록시 서버에 의해 저장되어 재사용될 수 있는지를 나타내는 속성
![image](https://github.com/user-attachments/assets/c8b2aaca-e599-4dd7-bd77-dde8d648638b)

**Reply by jongmee** [reply]
PUT, PATCH는 캐시가 안되지 않나요 ?? 👀 GET, HEAD만 [Cacheable](https://developer.mozilla.org/ko/docs/Glossary/Cacheable) 아닌가,,??

아닝가..?

**Reply by linirini** [reply]
[RFC 문서](https://datatracker.ietf.org/doc/html/rfc7231#section-4)에 따르면 일반적으로 현재 상태나 권위 있는 응답(주로 요청된 리소스의 최신 상태나 정확한 정보를 보장하는 응답)에 의존하지 않는 안전한 메서드는 캐시 가능으로 정의됩니다. 이 사양은 GET, HEAD, POST를 캐시 가능으로 정의하지만, 대부분의 캐시 구현은 GET과 HEAD만을 지원합니다.
[RFC 5789](https://www.rfc-editor.org/rfc/rfc5789)에 따르면 PATCH도 특정 조건을 만족할 때에는 캐싱이 가능하다고 합니다.
  - 응답에 명시적인 신선도 정보가 포함되어야 합니다.
      예: Expires 헤더 또는 Cache-Control: max-age 지시어.
  - 응답에 포함된 Content-Location 헤더가 요청된 URI(Request-URI)와 일치해야 합니다.
      이는 응답 본문이 리소스의 표현(representation)임을 나타냅니다.
POST와 PATCH도 가능은 하지만 거의 안쓴다고 생각하면 되지 않을까 싶습니다!

**Reply by jongmee** [reply]
오 그렇군요?! 감사합니다 🙇🏻‍♀️

**Comment by linirini** [comment]
메서드 | 역할 | 특성
-- | -- | --
GET | 데이터 조회 | 안전함, 멱등
POST | 데이터 생성 | 안전하지 않음, 멱등하지 않음
PUT | 데이터 전체 수정 | 멱등
PATCH | 데이터 일부 수정 | 멱등하지 않을 수도 있음
DELETE | 데이터 삭제 | 멱등
HEAD | 응답 헤더 조회 | 안전함, 멱등
OPTIONS | 지원하는 메서드 조회 | 안전함, 멱등
TRACE | 네트워크 경로 추적 | 안전함, 멱등 (보통 비활성화)
CONNECT | 터널링 | 보통 프록시에 사용
